File: Server~/src/tools/waitForUITool.ts
import * as z from 'zod';
import { Logger } from '../utils/logger.js';
import { McpUnity } from '../unity/mcpUnity.js';
import { McpServer } from '@modelcontextprotocol/sdk/server/mcp.js';
import { McpUnityError, ErrorType } from '../utils/errors.js';
import { CallToolResult } from '@modelcontextprotocol/sdk/types.js';

const toolName = 'wait_for_ui';
const toolDescription =
  "Polls for a GameObject by hierarchy path until it appears, is active, and (optionally) is interactable, " +
  "or a timeout elapses. Works in both Edit Mode and Play Mode. Useful as a synchronization primitive between " +
  "UI clicks — wait for the next FTUE step to render before acting on it. Default timeoutMs=5000, max 60000.";

// Custom timeout buffer: Unity-side wait can run up to MAX_WAIT_MS; the WebSocket request
// must outlive that, so we add 5s of headroom for transport overhead.
const MAX_WAIT_MS = 60_000;
const TRANSPORT_BUFFER_MS = 5_000;

const paramsSchema = z.object({
  objectPath: z
    .string()
    .min(1)
    .describe("Hierarchy path or leaf name of the GameObject to wait for, e.g. 'Canvas/Panel/StartButton'."),
  timeoutMs: z
    .number()
    .int()
    .min(0)
    .max(MAX_WAIT_MS)
    .optional()
    .default(5000)
    .describe(`How long to poll before giving up, in milliseconds. Default 5000, max ${MAX_WAIT_MS}.`),
  requireInteractable: z
    .boolean()
    .optional()
    .default(true)
    .describe(
      "If true (default), require the GameObject to be active AND not blocked by a CanvasGroup AND its " +
      "Selectable (if any) to be interactable. If false, only checks active state."
    ),
});

export function registerWaitForUITool(server: McpServer, mcpUnity: McpUnity, logger: Logger) {
  logger.info(`Registering tool: ${toolName}`);

  server.tool(
    toolName,
    toolDescription,
    paramsSchema.shape,
    async (params: any) => {
      try {
        logger.info(`Executing tool: ${toolName}`, params);
        const result = await toolHandler(mcpUnity, params);
        logger.info(`Tool execution successful: ${toolName}`);
        return result;
      } catch (error) {
        logger.error(`Tool execution failed: ${toolName}`, error);
        throw error;
      }
    }
  );
}

async function toolHandler(
  mcpUnity: McpUnity,
  params: z.infer<typeof paramsSchema>
): Promise<CallToolResult> {
  const timeoutMs = params.timeoutMs ?? 5000;
  const requireInteractable = params.requireInteractable ?? true;

  // Override the default 30s WebSocket request timeout so it can outlast the Unity-side wait.
  const requestTimeout = timeoutMs + TRANSPORT_BUFFER_MS;

  const response = await mcpUnity.sendRequest(
    {
      method: toolName,
      params: {
        objectPath: params.objectPath,
        timeoutMs,
        requireInteractable,
      },
    },
    { timeout: requestTimeout }
  );

  if (!response.success) {
    throw new McpUnityError(
      ErrorType.TOOL_EXECUTION,
      response.message || `wait_for_ui failed for '${params.objectPath}'`
    );
  }

  return {
    content: [
      {
        type: 'text',
        text: response.message,
      },
      {
        type: 'text',
        text: JSON.stringify(
          {
            objectPath: response.objectPath,
            elapsedMs: response.elapsedMs,
          },
          null,
          2
        ),
      },
    ],
  };
}

File: Editor/Tools/WaitForUITool.cs
using System;
using System.Threading.Tasks;
using McpUnity.Unity;
using McpUnity.Utils;
using Newtonsoft.Json.Linq;
using UnityEditor;
using UnityEngine;
using UnityEngine.UI;

namespace McpUnity.Tools
{
    /// <summary>
    /// Tool that polls until a target GameObject appears, is active, and (optionally) interactable.
    /// Useful as a synchronization primitive between ui_click_gameobject calls — wait for the next FTUE
    /// step to render before clicking it. Polling runs on EditorApplication.update (frame-driven).
    /// </summary>
    public class WaitForUITool : McpToolBase
    {
        private const int DefaultTimeoutMs = 5000;
        private const int MaxTimeoutMs = 60000;

        public WaitForUITool()
        {
            Name = "wait_for_ui";
            Description =
                "Polls for a GameObject by hierarchy path until it appears and (optionally) is interactable, " +
                "or a timeout elapses. Works in both Edit Mode and Play Mode (hierarchy traversal is mode-agnostic). " +
                "Returns success with elapsedMs when found, or 'timeout' on failure. Default timeout 5000 ms, max 60000 ms.";
            IsAsync = true;
        }

        public override void ExecuteAsync(JObject parameters, TaskCompletionSource<JObject> tcs)
        {
            string objectPath = parameters?["objectPath"]?.ToObject<string>();
            int timeoutMs = parameters?["timeoutMs"]?.ToObject<int?>() ?? DefaultTimeoutMs;
            bool requireInteractable = parameters?["requireInteractable"]?.ToObject<bool?>() ?? true;

            if (string.IsNullOrEmpty(objectPath))
            {
                tcs.TrySetResult(McpUnitySocketHandler.CreateErrorResponse(
                    "Missing required parameter: objectPath",
                    "validation_error"));
                return;
            }

            timeoutMs = Mathf.Clamp(timeoutMs, 0, MaxTimeoutMs);

            // Fast path: target may already be ready before the first editor update tick.
            if (TryMatch(objectPath, requireInteractable, out _))
            {
                tcs.TrySetResult(BuildSuccess(objectPath, 0L));
                return;
            }

            double startTime = EditorApplication.timeSinceStartup;
            EditorApplication.CallbackFunction tick = null;

            tick = () =>
            {
                try
                {
                    long elapsedMs = (long)((EditorApplication.timeSinceStartup - startTime) * 1000.0);

                    if (TryMatch(objectPath, requireInteractable, out _))
                    {
                        EditorApplication.update -= tick;
                        tcs.TrySetResult(BuildSuccess(objectPath, elapsedMs));
                        return;
                    }

                    if (elapsedMs >= timeoutMs)
                    {
                        EditorApplication.update -= tick;
                        tcs.TrySetResult(McpUnitySocketHandler.CreateErrorResponse(
                            $"Timed out after {elapsedMs} ms waiting for '{objectPath}' " +
                            $"(requireInteractable={requireInteractable}).",
                            "timeout"));
                    }
                }
                catch (Exception ex)
                {
                    EditorApplication.update -= tick;
                    McpLogger.LogError($"[MCP Unity] wait_for_ui tick failed: {ex.Message}\n{ex.StackTrace}");
                    tcs.TrySetResult(McpUnitySocketHandler.CreateErrorResponse(
                        $"Internal error during wait: {ex.Message}",
                        "internal_error"));
                }
            };

            EditorApplication.update += tick;
        }

        private static bool TryMatch(string objectPath, bool requireInteractable, out GameObject found)
        {
            found = GameObjectResolver.FindInLoadedScenes(objectPath);
            if (found == null) return false;
            if (!found.activeInHierarchy) return false;
            if (!requireInteractable) return true;
            return IsInteractable(found);
        }

        /// <summary>
        /// True when no CanvasGroup in the chain blocks raycasts/interactivity, and the on-object Selectable
        /// (if any) is interactable.
        /// </summary>
        private static bool IsInteractable(GameObject go)
        {
            Transform current = go.transform;
            while (current != null)
            {
                CanvasGroup group = current.GetComponent<CanvasGroup>();
                if (group != null)
                {
                    if (!group.interactable || !group.blocksRaycasts) return false;
                    if (group.ignoreParentGroups) break;
                }
                current = current.parent;
            }

            Selectable selectable = go.GetComponent<Selectable>();
            return selectable == null || selectable.IsInteractable();
        }

        private static JObject BuildSuccess(string objectPath, long elapsedMs)
        {
            return new JObject
            {
                ["success"] = true,
                ["type"] = "text",
                ["message"] = $"'{objectPath}' is ready after {elapsedMs} ms.",
                ["objectPath"] = objectPath,
                ["elapsedMs"] = elapsedMs
            };
        }
    }
}
